nltk/tag/mallet.py

```diff
@@ -0,0 +1,249 @@
+# -*- coding: utf-8 -*-
+# Natural Language Toolkit: Interface to the Stanford NER-tagger
+#
+# Copyright (C) 2001-2015 NLTK Project
+# Author: Long Duong <longdt219@gmail.com>
+# URL: <http://nltk.org/>
+# For license information, see LICENSE.TXT
+
+"""
+A module for interfacing with the Mallet CRF taggers.
+"""
+
+import os
+import tempfile
+from subprocess import PIPE
+import unicodedata
+import re 
+
+from nltk.internals import find_jar, config_java, java, _java_options
+from nltk.tag.api import TaggerI
+
+_mallet_url = 'http://mallet.cs.umass.edu/grmm/'
+
+class MalletTagger(TaggerI):
+    """
+    An interface to Mallet CRF taggers.
+    
+    >>> from nltk.tag.mallet import MalletTagger
+    >>> mt = MalletTagger('/Users/HeroAthen/Dropbox/Work/nltk/temp/mallet-2.0.7/dist/mallet.jar', 
+    ... '/Users/HeroAthen/Dropbox/Work/nltk/temp/mallet-2.0.7/dist/mallet-deps.jar')
+ 
+    >>> train_data = [[('University','Noun'), ('is','Verb'), ('a','Det'), ('good','Adj'), ('place','Noun')],
+    ... [('dog','Noun'),('eat','Verb'),('meat','Noun')]]
+    
+    >>> mt.train(train_data,'model.crf.tagger')
+    >>> mt.tag_sents([['dog','is','good'], ['Cat','eat','meat']])
+    [[('dog', 'Noun'), ('is', 'Verb'), ('good', 'Noun')], [('Cat', 'Noun'), ('eat', 'Verb'), ('meat', 'Noun')]]
+    
+    >>> gold_sentences = [[('dog','Noun'),('is','Verb'),('good','Adj')] , [('Cat','Noun'),('eat','Verb'), ('meat','Noun')]] 
+    >>> mt.evaluate(gold_sentences) - 0.83333 < 0.0001
+    True
+    
+    Test the function of finding jar file in the environment variable and setting learned model file  
+    >>> mt = MalletTagger() 
+    >>> mt.set_model_file('model.crf.tagger')
+    >>> mt.evaluate(gold_sentences) - 0.83333 < 0.0001
+    True 
+    
+    """
+    
+    _JAR = 'mallet.jar'
+    _JAR_DEPS = 'mallet-deps.jar'
+
+    def __init__(self, path_to_mallet_jar=None, path_to_mallet_deps_jar = None, verbose=False, java_options='-mx1000m'):
+
+        self._mallet_jar = find_jar(
+                self._JAR, path_to_mallet_jar,
+                env_vars=('MALLET',), searchpath=(), url=_mallet_url,
+                verbose=verbose)
+        
+        self._mallet_dep_jar = find_jar(
+                self._JAR_DEPS, path_to_mallet_deps_jar,
+                env_vars=('MALLET_DEPS',), searchpath=(), url=_mallet_url,
+                verbose=verbose) 
+        
+        self.java_options = java_options
+        self._model_file = ''
+    
+    def set_model_file(self, model_file):
+        self._model_file = model_file
+            
+    def _get_features(self,data):
+        """
+        Extract basic features about this word including 
+             - Current Word 
+             - Is Capitalized ?
+             - Has Punctuation ?
+             - Has Number ?
+             - Suffixes up to length 3 
+        :return : a string which contains the features   
```

Comment:
Is it an example Mallet wrapper for POS tagging? If so, where did the existing features came from? It could make sense to use features from some published paper, e.g. from http://www.inf.ed.ac.uk/resources/nlp/local_doc/mxpost_doc.pdf, with a reference. 

Alternatively, if it is not just an educational example, and if it is intended to be a production-quality POS tagger, we can make model use features which gives good results on a large treebank. 81% accuracy on POS tagging is far from the state of the arts; it may be explained by a small training data size (300/214 sentences is not a large treebank), but if so features shouldn't be chosen based on what gives best results on this dataset.